#2 - Add getting data from GitHub api

orlowski11 opened the pull request:
Added:

- services, job, controller for fetching data of organization's repositories, workflow runs and workflow jobs
- handling rate limit
- tests

orlowski11 committed:
#38 - Create models and migrations for data from GitHub api (#42)

* added models, migrations, dtos and factories

* cleanup

* Update app/DTO/WorkflowJobDTO.php

Co-authored-by: Kamil Piech <kamil.piech@blumilk.pl>

---------

Co-authored-by: Kamil Piech <kamil.piech@blumilk.pl>

orlowski11 committed:
Merge branch 'main' of https://github.com/blumilksoftware/gha-analyzer into #2-add-getting-data-from-github-api

orlowski11 committed:
#39 - Add fetching repositories (#53)

* added fetching repositories service and tests

* changed repositories to collection & added timestamps

* fixed organization id in creating dto

* removed throwing exception on empty collection

orlowski11 committed:
#40 - Add fetching workflow runs (#54)

added fetching workflow runs and tests

orlowski11 committed:
#41 - Add fetching workflow jobs (#55)

* added services and tests for fetching workflow jobs, getting runner data from labels, calculating job time

* added checking if jobs are already fetched before sending request

* fixes & carbon for calculating job time

orlowski11 committed:
- Fetching data from api async (#56)

* added route and controller method for fetching data

* changed to asynchrous fetching data

* removed unnecessary log

orlowski11 committed:
cleanup

orlowski11 committed:
added redis variables to env.ci

orlowski11 committed:
added redis to workflows

kamilpiech97 commented on .env.ci:
```diff
@@ -23,3 +23,8 @@ SESSION_LIFETIME=120
 
 GITHUB_WEBHOOK_SECRET=123
+
+REDIS_CLIENT=phpredis
```
not `redis`?

kamilpiech97 commented on .github/workflows/test-and-lint-php.yml:
```diff
@@ -33,6 +33,11 @@ jobs:
         ports:
           - 5432:5432
 
+      redis:
+        image: redis:7.2.4-alpine3.19
+        ports:
+          - 6379:6379
+
```
I think, on CI its not required. U can use sync mode.

kamilpiech97 commented on app/Http/Controllers/GithubController.php:
```diff
@@ -40,4 +45,19 @@ public function callback(): RedirectResponse
         return redirect("/");
     }
+
+    public function fetchData($organizationId): RedirectResponse
+    {
+        $userId = Auth::user()->id;
+
+        FetchDataFromApi::dispatch(
+            (int)$organizationId,
+            $githubConnector = new GithubConnector(),
+            new FetchRepositoriesService($githubConnector, $userId),
+            new FetchWorkflowRunsService($githubConnector, $userId),
+            new FetchWorkflowJobsService($githubConnector, $userId),
+        );
+
+        return redirect()->back();
+    }
```
Maybe for fetching data nice to be pack that in [queue](https://laravel.com/docs/11.x/queues)?

kamilpiech97 commented on database/factories/WorkflowJobFactory.php:
```diff
@@ -0,0 +1,32 @@
+{
+    public function definition(): array
+    {
+        $osArray = ["ubuntu", "windows"];
```
what about `apple`

AmonDeShir committed:
remove redis from workflow

AmonDeShir committed:
remove redis from env.ci

AmonDeShir committed:
add apple to job factory

AmonDeShir committed:
split FetchDataFromApi into multiple jobs

- Fixed a bug where AssignUserToOrganizationsService failed if the user organization did not exist
- Split FetchDataFromApi.php into multiple jobs

AmonDeShir committed:
fix linter errors

kamilpiech97 commented on database/factories/WorkflowJobFactory.php:
```diff
@@ -0,0 +1,36 @@
+            ["standard", "4", "8", "16", "32", "64"],
+            ["standard", "4", "8", "16", "32", "64"],
+            ["standard", "12"],
+        ];
```
Lets use values from config.

kamilpiech97 commented on app/Http/Controllers/GithubController.php:
```diff
@@ -40,4 +43,34 @@ public function callback(): RedirectResponse
+        $batch = Bus::findBatch($batchId);
+
+        if ($batch === null) {
+            return \response()->json(["message" => "Batch not found"], 404);
+        }
+
+        if ($batch->cancelled()) {
+            return \response()->json(["message" => "There was an error, please try again latter."], 500);
```
```suggestion
            return response()->json(["message" => "Batch not found"], 404);
        }

        if ($batch->cancelled()) {
            return response()->json(["message" => "There was an error, please try again latter."], 500);
```

kamilpiech97 replied:
Please be consistent

kamilpiech97 commented on app/Http/Integrations/GithubConnector.php:
```diff
@@ -23,4 +31,27 @@ protected function defaultHeaders(): array
+
+    protected function handleTooManyAttempts(Response $response, Limit $limit): void
+    {
+        if ($response->status() !== 429 && $response->status() !== 403) {
+            return;
+        }
```
```suggestion
        if (in_array($response->status(), [Status::HTTP_TOO_MANY_REQUESTS, Status::HTTP_FORBIDDEN]) {
            return;
        }
```

Lets use `Symfony\Component\HttpFoundation\Response` codes.

AmonDeShir committed:
use pricing values from config

AmonDeShir committed:
fix inconsistent response() usage

AmonDeShir committed:
use Symfony HTTP status codes

AmonDeShir committed:
use strict comparison in in_array()
